### PR TITLE
1.2 keyword references

### PR DESCRIPTION
**Pull request by ngilain**
[OO-FEATURE] Refacto Media usage tracking
https://github.com/open-orchestra/open-orchestra-mongo-libs/pull/50
https://github.com/open-orchestra/open-orchestra-model-interface/pull/225
https://github.com/open-orchestra/open-orchestra-model-bundle/pull/643
https://github.com/open-orchestra/open-orchestra-media-bundle/pull/233
https://github.com/open-orchestra/open-orchestra-media-admin-bundle/pull/282
https://github.com/open-orchestra/open-orchestra-cms-bundle/pull/1862
https://github.com/open-orchestra/open-orchestra-bbcode-bundle/pull/25
https://github.com/open-orchestra/open-orchestra/pull/1027

### CODE REVIEW

File: MediaAdmin/Reference/Strategies/AbstractMediaReferenceStrategy.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        return $references;
+    }
+
+    /**
+     * @param string $str
+     * @param array  $references
+     *
+     * @return array
+     */
+    protected function extractMediaFromBBCode($str, array $references)
```
Comment (veolys-nanne) [inline MediaAdmin/Reference/Strategies/AbstractMediaReferenceStrategy.php]: This method don't really need $references as parameter. Merge should be done in the calling method.

File: MediaAdmin/Reference/Strategies/MediaInNodeReferenceStrategy.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        return ($entity instanceof ReadNodeInterface);
+    }
+
+    /**
+     * @param mixed $entity
+     */
+    public function addReferencesToEntity($entity)
+    {
+        if ($this->support($entity)) {
```
Comment (veolys-nanne) [inline MediaAdmin/Reference/Strategies/MediaInNodeReferenceStrategy.php:29]: In others strategy + manager patterns, support is done in the manager loop, not directly in the strategies.

File: MediaAdmin/Reference/Strategies/AbstractMediaReferenceStrategy.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+abstract class AbstractMediaReferenceStrategy
+{
+    protected $mediaRepository;
+    protected $bbcodeParser;
+
+    /**
+     * MediaRepositoryInterface $mediaRepository
+     * BBcodeParser             $bbcodeParser
+     */
+    public function __construct(MediaRepositoryInterface $mediaRepository, BBcodeParser $bbcodeParser)
```
Comment (alavieille) [inline MediaAdmin/Reference/Strategies/AbstractMediaReferenceStrategy.php]: use BBcodeParserInterface

**Comment by alavieille** [comment]
can you add OO tag

File: MediaAdmin/Reference/Strategies/MediaInNodeReferenceStrategy.php (hunk 2 above)
Comment (ngilain) [reply MediaAdmin/Reference/Strategies/MediaInNodeReferenceStrategy.php:29]: @itkg-nanne : Yes but as discussed with @alavieille it's a fault that should be corrected on our other managers. According to the kind of strategy, giving an argument not supported can lead to something going from nothing particular to a fatal error like on these strategies.
Each method of a class (including the strategies ones) should always check by themselves that they support the parameters, they can't rely on an external source to do it.

Symfony voter are for instance coded that way.

File: MediaAdmin/Reference/Strategies/AbstractMediaReferenceStrategy.php (hunk 1 above)
Comment (ngilain) [reply MediaAdmin/Reference/Strategies/AbstractMediaReferenceStrategy.php]: @itkg-nanne : you're right, I correct this one and other method of this kind

**Comment by ngilain** [comment]
@alavieille , @itkg-nanne : update done

File: MediaAdmin/Reference/Strategies/MediaInNodeReferenceStrategy.php (hunk 2 above)
Comment (veolys-nanne) [reply MediaAdmin/Reference/Strategies/MediaInNodeReferenceStrategy.php:29]: :+1:

**Comment by veolys-nanne** [comment]
:+1:

**Comment by ngilain** [comment]
@alavieille : Cache invalidation after media update recoded after a rebase on your work

**Comment by alavieille** [comment]
:+1:

**Pull request merged**